File: pre_stack.cpp
#include <exception>
#include <iostream>
#include <string>
using namespace std;

class Stack
{
private:
	const int maxSize;
	char *stack;
	int pos;

public:
	class Full: public exception
	{
	public:
		virtual const char* what() const throw()
  		{
    		return "Stack is full";
  		}
	};

	class Empty: public exception
	{
	public:
		virtual const char* what() const throw()
  		{
    		return "Stack is empty";
  		}
	};

	/** Initialise the stack, maxSize is the maximum number of
		values that can be stored in the stack at any time */
	Stack( const int maxSize ) : maxSize(maxSize)
	{
		pos = 0;
		stack = new char[ maxSize ];
	}	

	~Stack()
	{
		delete [] stack;
	}

	/** Returns the number of values currently stored in the 
		stack */
	int num_items()
	{
		// REPLACE ME
		return 0;
	}

	/** Add value to the top of the stack, raises Stack::Full 
		exception if stack is full */
	void push( char value )
	{
		if( num_items() < maxSize )
		{
			stack[ pos ] = value;
			pos += 1;
		}
		else
		{
			throw Full();
		}
	}

	/** Returns the value currently stored at the top of the 
		stack, raises Stack::Empty exception if stack is 
		empty */
	char top()
	{
		// COMPLETE ME
	}

	/** Removes and returns the value currently stored at the
		top of the stack, raises Stack::Empty exception if stack
		is empty */
	char pop()
	{
		// COMPLETE ME
	}
};

int main()
{
	string testvalues = "abcde";

	Stack s( testvalues.length() );

	// === pushing test ======
	for( int i=0; i<testvalues.length(); ++i )
	{
		char c = testvalues[i];
		s.push( testvalues[i] );

		cout << "Pushing " << c << endl;

		if( s.top() != c )
		{
			cerr << "Error - last thing pushed was " << c << " but top of stack contains " << s.top() << endl;
		}

		if( s.num_items() != i+1 )
		{
			cerr << "Error - pushed " << i+1 << " values but stack reports size of " << s.num_items() << endl;
		}
	}

	// === is full test ======
	try
	{
		s.push( 'f' );
		cerr << "Error - tried to push to a full stack but no exception" << endl;
	}
	catch( Stack::Full& error )
	{}

	// === pop test ======
	for( int i=testvalues.length()-1; i>=0; --i )
	{
		char c = testvalues[i];
		char val = s.pop();

		cout << "Popping " << val << endl;

		if( val != c )
		{
			cerr << "Error - wrong value was popped from the stack, expecting " << c << " but got " << val << endl;
		}
	
		if( s.num_items() != i )
		{
			cerr << "Error - stack should have " << i << " values but claims it has " << s.num_items() << " values" << endl;
		}
	}

	// === empty test ======
	try
	{
		s.pop();
		cerr << "Error - tried to pop from an empty stack but no exception" << endl;
	}
	catch( Stack::Empty& error )
	{}

	try
	{
		s.top();
		cerr << "Error - tried to get top off an empty stack but no exception" << endl;
	}
	catch( Stack::Empty& error )
	{}

	return 0;
}

File: pre_stack.py
class Stack:
	class Full(Exception):
		pass
	class Empty(Exception):
		pass

	def __init__( self, maxSize ):
		""" Initialise the stack, maxSize is the maximum number of
			values that can be stored in the stack at any time """
		self.__maxSize = maxSize
		self.__stack = [None]*self.__maxSize
		self.__pos = 0

	def push( self, value ):
		""" Add value to the top of the stack, raises Stack.Full 
			exception if stack is full """
		if self.num_items() < self.__maxSize:
			self.__stack[self.__pos] = value
			self.__pos += 1
		else:
			raise self.Full( 'Can\'t push ' + str(value) )

	def num_items( self ):
		""" Returns the number of values currently stored in the 
			stack """
		
		# COMPLETE ME
		return 0

	def top( self ):
		""" Returns the value currently stored at the top of the 
			stack, raises Stack.Empty exception if stack is 
			empty """
		
		# COMPLETE ME
		pass

	def pop( self ):
		""" Removes and returns the value currently stored at the
			top of the stack, raises Stack.Empty exception if stack
			is empty """

		# COMPLETE ME
		pass

if __name__ == '__main__':
	import sys

	testvalues = 'abcde'
	s = Stack( len(testvalues) )
	
	# === pushing test ======
	for i in range(len(testvalues)):
		c = testvalues[i]
		s.push( c )

		print( 'Pushing %s' % c )

		if s.top() != c:
			print( 'Error - last thing pushed was %s but top of the stack contains %s' % (c, s.top()) )


		if s.num_items() != i+1:
			print( 'Error - pushed %d values but stack reports size of %d' % (i+1,s.num_items()) )

	# === is full test ======
	try:
		s.push( 'f' )
		print( 'Error - tried to push to a full stack but no exception' )
	except Stack.Full:
		pass

	# === pop test ======
	for i in range(len(testvalues)-1,-1,-1):
		c = testvalues[i]
		val = s.pop()

		print( 'Popping %s' % val )

		if val != c:
			print( 'Error - wrong value was popped from the stack, expecting %s but got %s' % (c, val) )

		if s.num_items() != i:
			print( 'Error - stack should have %d values but claims it has %d values' % (i,s.num_items()) )

	# === empty test ====
	try:
		s.pop()
		print( 'Error - tried to pop from an empty stack but no exception' )
	except Stack.Empty:
		pass

	try:
		s.top()
		print( 'Error - tried to get top of an empty stack but no exception' )
	except Stack.Empty:
		pass